### ISSUE TITLE
[NodeJS 17 or later] “npm run dev”, “npm start”, and “npm run deploy” throw exception

### ISSUE THREAD

**Issue by esperecyan** [mozilla/hubs#4861]
**Description**
When I run `npm run deploy` on NodeJS17, it throws an exception.

**To Reproduce**
1. Install NodeJS 17
1. Clone https://github.com/mozilla/hubs
1. `git checkout hubs-cloud` (52f293ed7f65329386adc50cbd192807166c83a5)
1. `npm ci`
1. `npm run login`
1. `npm run deploy`
1. See error

> PS ○○○○○○\hubs> npm run deploy
> 
> \> hubs@0.0.1 deploy
> \> node -r @babel/register -r esm -r ./scripts/shim scripts/deploy.js
> 
> Browserslist: caniuse-lite is outdated. Please run:
> npx browserslist@latest --update-db
> Deploying to exr-hubs.com.
> &nbsp;&nbsp;/ Building Client.node:child_process:397
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ex = new Error('Command failed: ' + cmd + '\n' + stderr);
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;^
> 
> Error: Command failed: npm run build
> Browserslist: caniuse-lite is outdated. Please run:
> npx browserslist@latest --update-db
> node:internal/crypto/hash:67
> &nbsp;&nbsp;this[kHandle] = new _Hash(algorithm, xofLen);
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;^
> 
> Error: error:0308010C:digital envelope routines::unsupported
> &nbsp;&nbsp;&nbsp;&nbsp;at new Hash (node:internal/crypto/hash:67:19)
> &nbsp;&nbsp;&nbsp;&nbsp;at Object.createHash (node:crypto:130:10)
> &nbsp;&nbsp;&nbsp;&nbsp;at module.exports (○○○○○○\hubs\node_modules\webpack\lib\util\createHash.js:135:53)
> &nbsp;&nbsp;&nbsp;&nbsp;at NormalModule._initBuildHash (○○○○○○\hubs\node_modules\webpack\lib\NormalModule.js:417:16)
> &nbsp;&nbsp;&nbsp;&nbsp;at handleParseError (○○○○○○\hubs\node_modules\webpack\lib\NormalModule.js:471:10)
> &nbsp;&nbsp;&nbsp;&nbsp;at ○○○○○○\hubs\node_modules\webpack\lib\NormalModule.js:503:5
> &nbsp;&nbsp;&nbsp;&nbsp;at ○○○○○○\hubs\node_modules\webpack\lib\NormalModule.js:358:12
> &nbsp;&nbsp;&nbsp;&nbsp;at ○○○○○○\hubs\node_modules\loader-runner\lib\LoaderRunner.js:373:3
> &nbsp;&nbsp;&nbsp;&nbsp;at iterateNormalLoaders (○○○○○○\hubs\node_modules\loader-runner\lib\LoaderRunner.js:214:10)
> &nbsp;&nbsp;&nbsp;&nbsp;at iterateNormalLoaders (○○○○○○\hubs\node_modules\loader-runner\lib\LoaderRunner.js:221:10)
> &nbsp;&nbsp;&nbsp;&nbsp;at ○○○○○○\hubs\node_modules\loader-runner\lib\LoaderRunner.js:236:3
> &nbsp;&nbsp;&nbsp;&nbsp;at context.callback (○○○○○○\hubs\node_modules\loader-runner\lib\LoaderRunner.js:111:13)
> &nbsp;&nbsp;&nbsp;&nbsp;at ○○○○○○\hubs\node_modules\babel-loader\lib\index.js:51:71
> &nbsp;&nbsp;&nbsp;&nbsp;at processTicksAndRejections (node:internal/process/task_queues:96:5) {
> &nbsp;&nbsp;opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
> &nbsp;&nbsp;library: 'digital envelope routines',
> &nbsp;&nbsp;reason: 'unsupported',
> &nbsp;&nbsp;code: 'ERR_OSSL_EVP_UNSUPPORTED'
> }
> 
> Node.js v17.1.0
> 
> &nbsp;&nbsp;&nbsp;&nbsp;at ChildProcess.exithandler (node:child_process:397:12)
> &nbsp;&nbsp;&nbsp;&nbsp;at ChildProcess.emit (node:events:390:28)
> &nbsp;&nbsp;&nbsp;&nbsp;at ChildProcess.emit (node:domain:475:12)
> &nbsp;&nbsp;&nbsp;&nbsp;at maybeClose (node:internal/child_process:1062:16)
> &nbsp;&nbsp;&nbsp;&nbsp;at Socket.<anonymous> (node:internal/child_process:448:11)
> &nbsp;&nbsp;&nbsp;&nbsp;at Socket.emit (node:events:390:28)
> &nbsp;&nbsp;&nbsp;&nbsp;at Socket.emit (node:domain:475:12)
> &nbsp;&nbsp;&nbsp;&nbsp;at Pipe.<anonymous> (node:net:687:12) {
> &nbsp;&nbsp;killed: false,
> &nbsp;&nbsp;code: 1,
> &nbsp;&nbsp;signal: null,
> &nbsp;&nbsp;cmd: 'npm run build'
> }
> 
> Node.js v17.1.0

**Expected behavior**
Deployment is completed successfully.

**Hardware** 
 - Device: Desktop
 - OS: Windows 10
 - Browser: Firefox

**Additional context**
I uninstalled NodeJS 17 and installed NodeJS 16, and the deployment completed successfully.

This may be caused by depending on an older version of webpack.
https://github.com/webpack/webpack/issues/14532

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/HUBSOLD-1171)

**Comment by esperecyan** [comment]
https://github.com/mozilla/hubs/blob/52f293ed7f65329386adc50cbd192807166c83a5/scripts/deploy.js#L44-L49
I was able to work around this problem by inserting `buildEnv.NODE_OPTIONS = "--openssl-legacy-provider";` into scripts/deploy.js.
```js
  buildEnv.BUILD_VERSION = `1.0.0.${version}`;
  buildEnv.ITA_SERVER = "";
  buildEnv.POSTGREST_SERVER = "";
  buildEnv.CONFIGURABLE_SERVICES = "janus-gateway,reticulum,hubs,spoke";
  // Workaround for npm run deploy failing on NodeJS 17
  // https://github.com/mozilla/hubs/issues/4861
  buildEnv.NODE_OPTIONS = "--openssl-legacy-provider";

  const env = Object.assign(process.env, buildEnv);
```

**Comment by esperecyan** [comment]
This problem also occurred with `npm run dev`, and similarly, I was able to work around it by setting the environment variable `NODE_OPTIONS` to `--openssl-legacy-provider`.

**Comment by msalafia** [comment]
Where should i define NODE_OPTIONS = "--openssl-legacy-provider" for `npm run start`?

**Comment by esperecyan** [comment]
@msalafia I wrote a script file that calls webpack-dev-server in the spawn() function to inject the environment variable.

**Comment by techcto** [comment]
I getting this `$ npm install --legacy-peer-deps
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'aframe@1.0.3',
npm WARN EBADENGINE   required: { node: '>= 4.6.0', npm: '^2.15.9' },
npm WARN EBADENGINE   current: { node: 'v16.14.2', npm: '8.5.0' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'networked-aframe@0.6.1',
npm WARN EBADENGINE   required: { node: '>= 7.4.0', npm: '^4.0.5' },
npm WARN EBADENGINE   current: { node: 'v16.14.2', npm: '8.5.0' }
npm WARN EBADENGINE }
npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
`

**Comment by techcto** [comment]
I am guessing these two sub repos need to be updated?  Any advice to building locally?

https://github.com/MozillaReality/aframe
https://github.com/MozillaReality/networked-aframe

Is anyone working on these two repos?

**Comment by netpro2k** [comment]
This should now be fixed with #5619 as we are now using node 16.16.0
